kernel/syscall.c

```diff
@@ -2656,6 +2656,86 @@ long myst_syscall_recvmsg(int sockfd, struct msghdr* msg, int flags)
     return ret;
 }
 
+long myst_syscall_sendmmsg(
+    int sockfd,
+    struct mmsghdr* msgvec,
+    unsigned int vlen,
+    int flags)
+{
+    long ret = 0;
+    myst_fdtable_t* fdtable = myst_fdtable_current();
+    myst_sockdev_t* sd;
+    myst_sock_t* sock;
+    unsigned int cnt;
+
+    ECHECK(myst_fdtable_get_sock(fdtable, sockfd, &sd, &sock));
+    if (!msgvec && vlen)
+        ERAISE(EFAULT);
+    for (cnt = 0; cnt < vlen; cnt++)
+    {
+        ret = (*sd->sd_sendmsg)(sd, sock, &msgvec[cnt].msg_hdr, flags);
+        if (ret < 0)
+            break;
+        msgvec[cnt].msg_len = ret;
+    }
+    // Only return err when zero msg was sent
+    ret = cnt ? (long)cnt : ret;
+
+done:
+    return ret;
+}
+
+long myst_syscall_recvmmsg(
+    int sockfd,
+    struct mmsghdr* msgvec,
+    unsigned int vlen,
+    int flags,
+    struct timespec* timeout)
+{
+    long ret = 0;
+    myst_fdtable_t* fdtable = myst_fdtable_current();
+    myst_sockdev_t* sd;
+    myst_sock_t* sock;
+    struct timespec start;
+    struct timespec now;
+    long expire;
+    unsigned int cnt;
+
+    ECHECK(myst_fdtable_get_sock(fdtable, sockfd, &sd, &sock));
+    if (!msgvec && vlen)
+        ERAISE(EFAULT);
+    if (timeout)
+    {
+        if (!is_timespec_valid(timeout))
+            ERAISE(EINVAL);
+
+        expire = get_nanos_from_timespec(timeout);
+        myst_syscall_clock_gettime(CLOCK_MONOTONIC, &start);
+    }
+    for (cnt = 0; cnt < vlen; cnt++)
+    {
+        ret = (*sd->sd_recvmsg)(
+            sd, sock, &msgvec[cnt].msg_hdr, flags & ~MSG_WAITFORONE);
```

Comment:
Please explain the setting and clearing of flag `MSG_WAITFORONE` in comments

Reply:
Fixed.